File: Sources/Runtime/Plugin/MelodyPluginRegistry.swift
import SwiftUI

/// Holds all registered ``MelodyPlugin`` instances and bulk-registers them on new VMs.
public final class MelodyPluginRegistry: @unchecked Sendable {
    private let plugins: [MelodyPlugin]

    public init(plugins: [MelodyPlugin] = []) {
        self.plugins = plugins
    }

    /// Register all plugin functions on the given VM.
    public func register(on vm: LuaVM) {
        for plugin in plugins {
            plugin.register(vm: vm)
        }
    }
}

// MARK: - Environment Key

private struct PluginRegistryKey: EnvironmentKey {
    static let defaultValue: MelodyPluginRegistry? = nil
}

extension EnvironmentValues {
    public var pluginRegistry: MelodyPluginRegistry? {
        get { self[PluginRegistryKey.self] }
        set { self[PluginRegistryKey.self] = newValue }
    }
}

File: Android/melody-runtime/src/main/java/com/melody/runtime/plugin/MelodyPluginRegistry.kt
package com.melody.runtime.plugin

import com.melody.runtime.engine.LuaVM

/** Holds all registered plugins and applies them to a LuaVM. */
class MelodyPluginRegistry(
    private val plugins: List<MelodyPlugin> = emptyList()
) {
    /** Register all plugin functions on the given VM. */
    fun register(vm: LuaVM) {
        for (plugin in plugins) {
            plugin.register(vm)
        }
    }
}
